### PR TITLE
BUG: arg validation in DTA/TDA/PA.take, DTI/TDI/PI.where

### PR DESCRIPTION
**Pull request by jbrockmendel**
We have a _lot_ of really similar validation/casting methods in these classes.  This is a step in smoothing out the differences between them so we can share them.

### CODE REVIEW

**Comment by jreback** [comment]
looks fine, pls rebase

**Comment by jbrockmendel** [comment]
rebased+green

File: pandas/tests/arrays/test_datetimelike.py [hunk 1]
```diff
@@ -508,6 +508,12 @@ def test_take_fill_valid(self, datetime_index, tz_naive_fixture):
             # require NaT, not iNaT, as it could be confused with an integer
             arr.take([-1, 1], allow_fill=True, fill_value=value)
 
+        value = np.timedelta64("NaT", "ns")
+        msg = f"'fill_value' should be a {self.dtype}. Got '{str(value)}'."
```
Comment (simonjayhawkins) [inline pandas/tests/arrays/test_datetimelike.py:512]: I _think_ normally shouldn't need str in f-string. could this be a numpy bug?

Comment (jbrockmendel) [reply pandas/tests/arrays/test_datetimelike.py:512]: i forget exactly which case broke without the extra `str`, but plausibly numpy

**Pull request merged**